From 082e1d3d2004a6bf61009647597d6f1cabac79d5 Mon Sep 17 00:00:00 2001
From: Nicolas Adenis-Lamarre <nicolas.adenis.lamarre@gmail.com>
Date: Mon, 24 Oct 2022 12:59:58 +0000
Subject: [PATCH] use real vendor/product id

the current way doesn't provide the real vendor/produc ids
that can be found via evtest.
This patch applies the same commands from the evtest source code.
It works for any pad, including bluetooth pads.

This patch was originally written for the batocera project at:
https://github.com/batocera-linux/batocera.linux/tree/master/board/batocera/patches/kodi-peripheral-joystick

Signed-off-by: Nicolas Adenis-Lamarre <nicolas.adenis.lamarre@gmail.com>
---
 src/api/udev/JoystickUdev.cpp | 19 ++++++++++---------
 1 file changed, 10 insertions(+), 9 deletions(-)

diff --git a/src/api/udev/JoystickUdev.cpp b/src/api/udev/JoystickUdev.cpp
index 3c0c6113..2ef41a2b 100644
--- a/src/api/udev/JoystickUdev.cpp
+++ b/src/api/udev/JoystickUdev.cpp
@@ -275,15 +275,16 @@ bool CJoystickUdev::GetProperties()
   }
   SetName(name);
 
-  // Don't worry about unref'ing the parent
-  struct udev_device* parent = udev_device_get_parent_with_subsystem_devtype(m_dev, "usb", "usb_device");
-
-  const char* buf;
-  if ((buf = udev_device_get_sysattr_value(parent, "idVendor")) != nullptr)
-    SetVendorID(strtol(buf, NULL, 16));
-
-  if ((buf = udev_device_get_sysattr_value(parent, "idProduct")) != nullptr)
-    SetProductID(strtol(buf, NULL, 16));
+  unsigned short id[4];
+  char val[16];
+  if(ioctl(m_fd, EVIOCGID, id) == 0) {
+    sprintf(val, "%x", id[ID_VENDOR]);
+    SetVendorID(strtol(val, NULL, 16));
+    esyslog("joystick information vendorid=%s for %s", val, m_path.c_str());
+    sprintf(val, "%x", id[ID_PRODUCT]);
+    SetProductID(strtol(val, NULL, 16));
+    esyslog("joystick information productid=%s for %s", val, m_path.c_str());
+  }
 
   struct stat st;
   if (fstat(m_fd, &st) < 0)
